### ISSUE TITLE
Issues with pillow and jupyterlab versions

### ISSUE THREAD

**Issue by carlhiggs** [healthysustainablecities/global-indicators#125]
Dependabot alerts have been generated for pillow and jupyterlab versions currently used in our Docker image.

https://github.com/global-healthy-liveable-cities/global-indicators/security/dependabot/docker/requirements.txt/pillow/open
https://github.com/global-healthy-liveable-cities/global-indicators/security/dependabot/docker/requirements.txt/jupyterlab/open

To resolve, it is recommended to use 
pillow>=8.3.2
jupyterlab>=2.2.10

Currently 
pillow == 8.2.*
jupyterlab == 2.1.*

Before updating to fixed versions in the above recommended ranges, it should be confirmed that our code remains stable

**Comment by gboeing** [comment]
> Before updating to fixed versions in the above recommended ranges, it should be confirmed that our code remains stable

I'm not sure we need to update this. Nearly any "frozen" environment will eventually demonstrate dependency vulnerabilities over the years. The whole environment is containerized so it's not like it's running un-sandboxed on someone's production server. I think we should keep it snapshotted like this as the reference environment, that is, the exact configuration of code and underlying software used to generate the analytics reported in the paper. Anyone else's thoughts?

**Comment by carlhiggs** [comment]
That does sound good to me -- I had a few goes re-building tweaking the requirements versions to get rid of conflicts and haven't yet had success.   It might be at some point we could go over the code with a comb to see if we are installing things that aren't used which complicate things (eg nodejs was listed in Docker requirements.txt and may not be used in this project?).  But perhaps its not the priority right now.  I'm happy to close this if this isn't an issue worth resolving.

**Comment by carlhiggs** [comment]
@gboeing In case its of use later, I was able to successfully build using the following items in requirements.txt which should address these warnings --- however I haven't confirmed that all our code runs as intended.  I agree with your point that changing software at this point is not a good idea, so I'll close this issue but leave this list in case its of use if this is re-approached at a later date.
```
bottleneck
contextily
cython
flake8
geoalchemy2 == 0.6.*
geopandas == 0.7.*
jupyterlab >= 2.2.1,<3.0
memory_profiler == 0.57.*
networkx == 2.5.*
numexpr == 2.7.*
numpy == 1.19.*
osmnx == 1.1.*
pandana == 0.6.*
pillow >= 8.3.2,<8.4
pip
psycopg2 == 2.8.*
pylint
pysal == 2.2.*
python == 3.7.*
rasterio == 1.2.*
scikit-learn == 0.23.*
scipy == 1.4.*
sphinx
sphinxcontrib-bibtex
sphinx_rtd_theme
urbanaccess == 0.2.*
xlrd
xlwt
yapf == 0.29.*
```

I noticed that contextily is actually installed using pip directly in the Dockerfile, which may be necessary but I suppose that one or the other listing (in requirements.txt for conda, or the call to pip) could be removed.

If we later re-factor our configuration code to be more centralised (ie. better integrate the 'pre-process' and process configurations), we could also likely do away with xlrd and xlwt which would simplify things too.

But yes, these are things for a future release!

(incidentally, I'm looking to re-factor the configuration code in the spatial indicators software developed for the Bangkok project which we adopted here for pre-processing to use YAML or TOML instead of a workbook of spreadsheets for configuration; so perhaps once that is done, and possible other refinements done in that project, we may later be able to integrate these things to simplify things here)
